x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsIndex.java

```diff
@@ -176,17 +191,50 @@ private static void addMetaData(Map<String, Object> mappingsAsMap, String analyt
         return value;
     }
 
-    public static void updateMappingsToDestIndex(Client client, DataFrameAnalyticsConfig analyticsConfig, GetIndexResponse getIndexResponse,
+    @SuppressWarnings("unchecked")
+    public static void updateMappingsToDestIndex(Client client, DataFrameAnalyticsConfig config, GetIndexResponse getIndexResponse,
                                                  ActionListener<AcknowledgedResponse> listener) {
         // We have validated the destination index should match a single index
         assert getIndexResponse.indices().length == 1;
 
-        Map<String, Object> addedMappings = Map.of(PROPERTIES, Map.of(ID_COPY, Map.of("type", "keyword")));
+        ActionListener<ImmutableOpenMap<String, MappingMetaData>> mappingsListener = ActionListener.wrap(
+            mappings -> {
+                // Fetch merged mappings from source indices
+                Map<String, Object> sourceMappingsAsMap = mappings.valuesIt().next().sourceAsMap();
+                Map<String, Object> sourcePropertiesAsMap =
+                    (Map<String, Object>)sourceMappingsAsMap.getOrDefault(PROPERTIES, Collections.emptyMap());
+
+                // Verify that the results field does not exist in the source indices
+                checkResultsFieldIsNotPresentInSourceIndex(config, sourcePropertiesAsMap);
+
+                // Determine mappings to be added to the destination index
+                Map<String, Object> properties = new HashMap<>();
+                addProperties(properties, config, Collections.unmodifiableMap(sourcePropertiesAsMap));
+                Map<String, Object> addedMappings = Map.of(PROPERTIES, properties);
+
+                // Add the mappings to the destination index
+                PutMappingRequest putMappingRequest =
+                    new PutMappingRequest(getIndexResponse.indices())
+                        .source(addedMappings);
+                ClientHelper.executeWithHeadersAsync(
+                    config.getHeaders(), ML_ORIGIN, client, PutMappingAction.INSTANCE, putMappingRequest, listener);
+            },
+            listener::onFailure
+        );
 
-        PutMappingRequest putMappingRequest = new PutMappingRequest(getIndexResponse.indices());
-        putMappingRequest.source(addedMappings);
-        ClientHelper.executeWithHeadersAsync(analyticsConfig.getHeaders(), ML_ORIGIN, client, PutMappingAction.INSTANCE,
-            putMappingRequest, listener);
+        MappingsMerger.mergeMappings(client, config.getHeaders(), config.getSource(), mappingsListener);
```

Comment:
I don't think we want to merge mappings when the dest index exists. The idea here is that we allow the flexibility for the user to create custom mappings.